synapse/app/synctl.py

```diff
@@ -190,7 +199,11 @@ def main():
         if start_stop_synapse:
             stop(pidfile, "synapse.app.homeserver")
 
-        # TODO: Wait for synapse to actually shutdown before starting it again
+    # Wait for synapse to actually shutdown before starting it again
+    if action == "restart":
+        write("Waiting to restart ...")
+        while os.path.exists(pidfile):
```

Comment:
I have 2 solutions for this either 
• I use psutil , the cross-platform library
• keep calling `os.kill(pid, 0)` until it exits 
   but will have have to keep track of OSErrors